[TP][DTensor Perf]Fix DTensor Spec hash

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107181
* #106524

https://github.com/pytorch/pytorch/pull/106524 gets merged so fast that we didn't figure out that we should hash both stride and dtype in DTensorSpec. This is a forward fix.

One analysis for why using just shape is not enough.
1. We use the hash value for sharding propogation cache. And the output sharding contains the stride, size of the output DTensor. If we don't consider stride, we will see errors. 
2. One reason can be found below:
```
OpSchema(func_schema=aten::t(Tensor(a) self) -> Tensor(a), args_schema=(DTensorSpec(mesh=DeviceMesh:([0, 1, 2, 3, 4, 5, 6, 7]), placements=(Shard(dim=0),), tensor_meta=TensorMetadata(shape=torch.Size([64, 128]), dtype=torch.float32, requires_grad=False, stride=(128, 1), memory_format=None, is_quantized=False, qparams={})),), kwargs_schema={}) 
```


```
OpSchema(func_schema=aten::t(Tensor(a) self) -> Tensor(a), args_schema=(DTensorSpec(mesh=DeviceMesh:([0, 1, 2, 3, 4, 5, 6, 7]), placements=(Shard(dim=0),), tensor_meta=TensorMetadata(shape=torch.Size([64, 128]), dtype=torch.float32, requires_grad=False, stride=(1, 64), memory_format=None, is_quantized=False, qparams={})),), kwargs_schema={})
```

The only difference between two op_schame is the tensor stride:
<img width="151" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/161335df-bdfb-47c5-ba79-82616d070d15">

that makes the transpose op generates wrong result and leads to the add_/addmm_ op failing with errors:
```
Traceback (most recent call last):
  File "/data/users/fduwjj/pytorch/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "/data/users/fduwjj/pytorch/benchmarks/distributed/tensor/tp_benchmark.py", line 210, in run_tp
    output.sum().backward()
  File "/data/users/fduwjj/pytorch/torch/_tensor.py", line 491, in backward
    torch.autograd.backward(
  File "/data/users/fduwjj/pytorch/torch/autograd/__init__.py", line 251, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/data/users/fduwjj/pytorch/torch/distributed/_tensor/api.py", line 252, in __torch_dispatch__
    return op_dispatch.operator_dispatch(
  File "/data/users/fduwjj/pytorch/torch/distributed/_tensor/dispatch.py", line 116, in operator_dispatch
    out, _, _ = _operator_dispatch(op_call, args, kwargs, sharding_propagator)
  File "/data/users/fduwjj/pytorch/torch/distributed/_tensor/dispatch.py", line 246, in _operator_dispatch
    local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
  File "/data/users/fduwjj/pytorch/torch/_ops.py", line 435, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: The size of tensor a (64) must match the size of tensor b (8) at non-singleton dimension 1
```

Same thing with dtype, if we are using DTensor in the environment of mixed precision, we will run into situations like this.

fduwjj committed:
Fix DTensor Spec hash

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -378,13 +378,20 @@ class DTensorSpec:
+                    self.mesh,
+                    self.placements,
+                    self.tensor_meta.shape,
+                    self.tensor_meta.dtype,
```
why we need to hash the dtype and stride?

Also if self.tensor_meta is None then we don't hash, although every dtensor spec should be populated with this field, it would be better if we raise a error if this becomes None

fduwjj replied:
Oh, because for the sharding prop, if we don't include stride, the transpose op will get wrong results, so that in a simple TP MLP example, we run into errors in the backward. That's why originally, we didn't figure out this in the non-periodic unit tests.

fduwjj replied:
> Also if self.tensor_meta is None then we don't hash

yes. let me update the PR with all local changes.

wanchaol replied:
I see, did we try to update both __hash__ and __eq__ and only compare with shape (i.e. deleting dtype/stride in __eq__ too) and whether that works?

fduwjj replied:
No, that will cause errors. Because, we need stride to differentiate between different op_schema, specifically, the DTensorSpec within args_schema.

wanchaol replied:
hmm I see, do you have some example about DTensorSpecs that have the same placement, shape, but different stride,  and they need different output sharding? I am curious to see why that would cause errors.

fduwjj committed:
Update on "[TP][DTensor Perf]Fix DTensor Spec hash"


https://github.com/pytorch/pytorch/pull/106524 gets merged so fast that we didn't figure out that we should hash both stride and dtype in DTensorSpec. This is a forward fix.



[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -378,11 +378,20 @@ class DTensorSpec:
+                (
+                    self.mesh,
+                    self.placements,
+                    self.tensor_meta.shape,
```
I guess we don't need dtype, shall we delete `dtype` in both hash and eq then?

fduwjj replied:
Do we expect DTensor not to be used in mixed precision scenario? because this can cause wrong result in the mixed precision situation.

wanchaol replied:
for mixed precision we still have the same sharding propagation rules, so dtype won't affect how we conduct sharding propagations IMO, so likely we don't need dtype here

fduwjj replied:
but the output sharding contains wrong dtype right?

XilunWu replied:
Correct me if I was wrong:
it seems the usage of prop cache becomes quite restrictive: only all attributes match then is there a cache hit. Probably we adjust the cached OutputSharding object's dtype before returning it?

wanchaol reviewed (approved):
lgtm, it would be great if we can do some analysis about why simply hashing shape would not work

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -f "the failing test is about quantization which does not use DTensor. Another one gets canceled."
